Added ability to use user config file.

omron93 opened the pull request:
As discussed in #137 

Added ability to use user config file.
- user is able to create or mount custom config file  in `/etc/mongod.conf` (for example from non-docker MongoDB deployments)
- this config file does not affect MONGODB_DATADIR and MONGODB_REPLICA_NAME

Cleaned up MONGODB env variables and mongod parameters.
- all env var declations were moved into one place
- removed PIDFILE option (mongod stores PID into mongod.lock file in dbpath)
- fixed double declaration of MONGODB_KEYFILE_PATH

Commit:
Added ability to use user config file.
Cleaned up MONGODB env variables and mongod parameters.

Fixed typo (noprellocj).

Removed unnecessary exports.

Added readonly for constants. Constants reorganization.

Fixed user keyFile detection and bash string format.

Update README.

php-coder commented on 2.4/root/usr/share/container-scripts/mongodb/common.sh:
```diff
@@ -253,11 +261,17 @@ function mongo_reset_passwords() {
+# add --keyFile argument to mongo_common_args
 function setup_keyfile() {
+  # If user specify keyFile in cofig file do not use generated keyFile
+  if [ grep -q "keyFile" ${MONGODB_CONFIG_PATH} ]; then
```
BTW, you could remove `[` here: `if grep -q "keyFile" ${MONGODB_CONFIG_PATH} ; then` should work the same.

omron93 replied:
Thanks.

php-coder commented on 2.6/test/run:
```diff
@@ -280,9 +280,53 @@ function run_change_password_test() {
+    echo """dbpath=/var/lib/mongodb/dbpath
+    unixSocketPrefix = /var/lib/mongodb
+    smallfiles = true
+    nopreallocj = true""" > $config_file
```
`nopreallocj` <-- is it a typo? (`j` at the end)

omron93 replied:
Yes. Fixed.

rhcarvalho commented on 2.4/root/usr/share/container-scripts/mongodb/common.sh:
```diff
@@ -4,15 +4,23 @@ set -o errexit
+
+# Constants used for waiting
+export MAX_ATTEMPTS=60
+export SLEEP_TIME=1
```
Why export these variables?

omron93 replied:
You are right. Exports are needed only for envsub, right?

So removed other unnecessary exports.

rhcarvalho replied:
Yeah, MAX_ATTEMPTS and SLEEP_TIME are globals only used in this file.

omron93 replied:
Also MONGODB_CONFIG_PATH and MONGODB_KEYFILE_PATH, or not?

/usr/share/container-scripts/mongodb/common.sh is sourced in every run or execed file and config file does not contain them... so it is necessary to export them.

rhcarvalho replied:
> Also MONGODB_CONFIG_PATH and MONGODB_KEYFILE_PATH, or not?

Here's where Bash haunts our souls again... :cry: 
I'm not sure if it is necessary to export in that case, I guess not necessary, need to try it out.

Here are the occurrences of MONGODB_CONFIG_PATH:

``` console
$ git grep -n MONGODB_CONFIG_PATH upstream/master -- 2.4
upstream/master:2.4/root/usr/bin/run-mongod:67:envsubst < ${CONTAINER_SCRIPTS_PATH}/mongodb.conf.template > $MONGODB_CONFIG_PATH
upstream/master:2.4/root/usr/bin/run-mongod:71:mongo_common_args="-f $MONGODB_CONFIG_PATH --oplogSize 64"
upstream/master:2.4/root/usr/share/container-scripts/mongodb/common.sh:11:export MONGODB_CONFIG_PATH=/etc/mongod.conf
upstream/master:2.4/root/usr/share/container-scripts/mongodb/initiate_replica.sh:72:mongod -f ${MONGODB_CONFIG_PATH} --shutdown &>/dev/null
```

Trying it out:

``` console
$ cat foo
#!/bin/bash

set -o errexit
set -o nounset
set -o pipefail

source bar

echo $V_EXPORTED
echo $V_NOT_EXPORTED
$ cat bar
#!/bin/bash

export V_EXPORTED="exported var"
V_NOT_EXPORTED="not exported var"
$ ./foo 
exported var
not exported var
```

Conclusion: no need to export MONGODB_CONFIG_PATH and MONGODB_KEYFILE_PATH, @omron93 agree?

BTW, MONGODB_KEYFILE_PATH is defined/exported both in `2.4/root/usr/bin/run-mongod:21` and `2.4/root/usr/share/container-scripts/mongodb/common.sh:13`, we should likely get rid of one of those lines. Notice that $HOME == /var/lib/mongodb.

omron93 replied:
> Conclusion: no need to export MONGODB_CONFIG_PATH and MONGODB_KEYFILE_PATH, 

Agree.

> BTW, MONGODB_KEYFILE_PATH is defined/exported both in 2.4/root/usr/bin/run-mongod:21 and 2.4/root/usr/share/container-scripts/mongodb/common.sh:13, we should likely get rid of one of those lines. Notice that $HOME == /var/lib/mongodb.

This is not true in this PR.

rhcarvalho commented on 2.4/root/usr/share/container-scripts/mongodb/common.sh:
```diff
@@ -8,8 +8,8 @@ set -o pipefail
-export MONGODB_CONFIG_PATH=/etc/mongod.conf
-export MONGODB_KEYFILE_PATH="${HOME}/keyfile"
+MONGODB_CONFIG_PATH=/etc/mongod.conf
+MONGODB_KEYFILE_PATH="${HOME}/keyfile"
```
How about organizing it such that exported and non-exported variables are separated?
I think it would be less puzzling for readers.

omron93 replied:
This sounds good.

rhcarvalho commented on 2.4/root/usr/share/container-scripts/mongodb/common.sh:
```diff
@@ -18,8 +18,8 @@ export MONGODB_SMALLFILES=${MONGODB_SMALLFILES:-true}
-export MAX_ATTEMPTS=60
-export SLEEP_TIME=1
+MAX_ATTEMPTS=60
+SLEEP_TIME=1
```
We may want to make these `readonly`, if they're really constants not intended to be tweaked.
Or if we want to make it possible to customize them, maybe something like `MAX_ATTEMPTS={MAX_ATTEMPTS:-60}`, WDYT?

Since we never tweaked these values (not possible with current implementation), it'd vote for `readonly`.

omron93 replied:
Agree. Added readonly.

rhcarvalho commented on 2.4/root/usr/bin/run-mongod:
```diff
@@ -6,20 +6,6 @@ set -o pipefail
-export MONGODB_QUIET=${MONGODB_QUIET:-true}
-
-
-export MONGODB_KEYFILE_SOURCE_PATH="/var/run/secrets/mongo/keyfile"
```
I can't find where else this is defined?

omron93 replied:
Nowhere. I can't fine where it is used?

So I removed it.

rhcarvalho replied:
Seems right to remove it. It was originally added together with the replication example in https://github.com/openshift/mongodb/pull/17, but later its usage was removed in https://github.com/openshift/mongodb/pull/45 and the export was left over. Thanks for spotting it!

rhcarvalho commented on 2.4/root/usr/bin/run-mongod:
```diff
@@ -6,20 +6,6 @@ set -o pipefail
-
-
-export MONGODB_KEYFILE_SOURCE_PATH="/var/run/secrets/mongo/keyfile"
-export MONGODB_KEYFILE_PATH="${HOME}/keyfile"
```
I see this was defined both in `common.sh` and `run-mongod`, removal looks good.

omron93 commented:
@rhcarvalho @php-coder Any other suggestions?

rhcarvalho commented on 2.4/root/usr/bin/run-mongod:
```diff
@@ -46,8 +32,8 @@ function cleanup() {
-  if [ -f "${MONGODB_PID_FILE}" ]; then
-    kill -2 $(cat ${MONGODB_PID_FILE})
+  if [[ -s ${MONGODB_DATADIR}/mongod.lock ]]; then
+    kill -2 $(cat ${MONGODB_DATADIR}/mongod.lock)
```
How is this change related to user ability to provide a config file?

omron93 replied:
If the image provides ability to mount config file scripts should avoid getting values from it. (I thought that OpenShift suggested not to do this to avoid bugs for some user weird config files)

MONGODB_PID_FILE is one example of it. The same value is stored in file `mongod.lock` in dbpath. So no reason to get value of MONGODB_PID_FILE from config file.

rhcarvalho replied:
Maybe there should be some thought about backwards-compat before moving on with this change?

Is MONGODB_PID_FILE being read from the config file?!

omron93 replied:
What "backward-compat" do you mean?

`MONGODB_PID_FILE` before this PR is specified in `/usr/share/container-scripts/mongodb/common.sh` and it is used to stop container. In this PR `MONGODB_PID_FILE` is not used in scripts, so no need have it present in scripts.
However if you know some usecase when MONGODB_PID_FILE should be created with default config file. It could be easily added back to config.tepmlate.

omron93 replied:
@rhcarvalho Do you know why `cleanup()` uses `kill` or `pkill` instead of `mongod --shutdown`?

rhcarvalho replied:
> Do you know why cleanup() uses kill or pkill instead of mongod --shutdown?

I don't know, probably using `--shutdown` would be preferred. But again, it's easier to review PRs that do one thing.
That change would be much easier to review in a PR of itself IMHO.

omron93 replied:
> it's easier to review PRs that do one thing.
> That change would be much easier to review in a PR of itself IMHO.

Yes. But it depends on graininess. If "every line change" is submited in separate PR, it really makes improvements harder and makes thinks slower.

However, there is a new PR #155

rhcarvalho commented on 2.4/root/usr/bin/run-mongod:
```diff
@@ -63,12 +49,15 @@ if [ $# -gt 0 ] && [ "$1" == "initiate" ]; then
 # Need to cache the container address for the cleanup
 cache_container_addr
-mongo_common_args="-f $MONGODB_CONFIG_PATH --oplogSize 64"
+mongo_common_args="-f $MONGODB_CONFIG_PATH --dbpath $MONGODB_DATADIR"
```
Why now we need to pass a custom dbpath?

omron93 replied:
Necessity to override user dbpath was removed in #157. So removed from this PR.

rhcarvalho commented on 2.4/test/run:
```diff
@@ -280,9 +280,53 @@ function run_change_password_test() {
+    local volume_dir
+    volume_dir=`mktemp -d --tmpdir mongodb-testdata.XXXXX`
+    config_file=$volume_dir/mongod.conf
+    echo """dbpath=/var/lib/mongodb/dbpath
```
Why the triple double quotes here?!

Looks like Python, but this is not Python :P

omron93 replied:
You are right. Fixed.

rhcarvalho commented on 2.4/root/usr/bin/run-mongod:
```diff
@@ -114,8 +103,7 @@ else
     unset_env_vars
-    mongod $mongo_common_args --replSet ${MONGODB_REPLICA_NAME} \
-      --keyFile ${MONGODB_KEYFILE_PATH} ${auth_args}
+    mongod $mongo_common_args --replSet ${MONGODB_REPLICA_NAME} ${auth_args}
```
I see this was already lacking proper quoting... but maybe it's a good time to add it?

`MONGODB_REPLICA_NAME` should surely be double-quoted, while `mongo_common_args` and `auth_args` we may want them to expand to multiple things... seems weird to have 2 vars that expand to arbitrary things.
Even then, maybe we can quote and treat them as arrays with the magical @ syntax...

omron93 replied:
I agree with quoting  `MONGODB_REPLICA_NAME`. I don't know why `auth_args` are separated (Also this is not necessary for MongoDB 2.6+ because `--keyFile` implies auth). So `--auth` could be simply added to `mongo_common_args` for MongoDB 2.4.

@rhcarvalho Are you fine with doing it in this PR?

omron93 replied:
New PR for this #156

rhcarvalho commented on 2.4/root/usr/share/container-scripts/mongodb/common.sh:
```diff
@@ -253,11 +261,17 @@ function mongo_reset_passwords() {
+  fi
   if [ -z "${MONGODB_KEYFILE_VALUE-}" ]; then
-    echo "ERROR: You have to provide the 'keyfile' value in $MONGODB_KEYFILE_VALUE"
+    echo "ERROR: You have to provide the 'keyfile' value in MONGODB_KEYFILE_VALUE"
```
Don't know why this is still showing up here, but it is already in master... could you please rebase your PR (and probably squash the commits)?

omron93 replied:
Rebased.

rhcarvalho commented on 2.4/root/usr/share/container-scripts/mongodb/common.sh:
```diff
@@ -253,11 +261,17 @@ function mongo_reset_passwords() {
 # mounts the secret files with 'too open' permissions.
+# add --keyFile argument to mongo_common_args
 function setup_keyfile() {
+  # If user specify keyFile in cofig file do not use generated keyFile
```
s/cofig/config

omron93 replied:
Fixed.

rhcarvalho commented on 2.4/root/usr/share/container-scripts/mongodb/common.sh:
```diff
@@ -253,11 +261,17 @@ function mongo_reset_passwords() {
+# add --keyFile argument to mongo_common_args
 function setup_keyfile() {
+  # If user specify keyFile in cofig file do not use generated keyFile
+  if grep -q "keyFile" ${MONGODB_CONFIG_PATH}; then
```
this is a very broad grep... I'd prefer something more specific to avoid matching the string "keyFile" in some arbitrary place of the config, like a value or comment, not a key.

omron93 replied:
Now matching only lines starting with `keyFile`. Should be valid for both MongoDB config file formats.

rhcarvalho commented on 2.4/root/usr/share/container-scripts/mongodb/mongodb.conf.template:
```diff
@@ -1,7 +1,6 @@
 # mongodb.conf
 
-port = 27017
-pidfilepath = ${MONGODB_PID_FILE}
```
Why are we removing MONGODB_PID_FILE? If we are to remove it, we should update the README / docs too (there might be OpenShift docs referring to it).

omron93 replied:
Same as in https://github.com/openshift/mongodb/pull/147#discussion_r61854214 MongoDB scripts won't use this variable. 

@rhcarvalho Who can use this? Is it somehow possible use this file with PID in other dependent applications in different containers?

What docs please? I can't find any...

rhcarvalho replied:
If it's just used internally then I see no problems with removing it.

omron93 commented on 2.4/root/usr/bin/run-mongod:
```diff
@@ -63,12 +49,15 @@ if [ $# -gt 0 ] && [ "$1" == "initiate" ]; then
 # Need to cache the container address for the cleanup
 cache_container_addr
-mongo_common_args="-f $MONGODB_CONFIG_PATH --oplogSize 64"
+mongo_common_args="-f $MONGODB_CONFIG_PATH --dbpath $MONGODB_DATADIR"
```
To avoid getting values from config file - https://github.com/openshift/mongodb/pull/147#discussion_r61854214

`dbpath` is one example and it is really necessary for MongoDB scripts. Also for better performance database should be stored in the docker volume. And user mounting config files from "classical" deployments don't have to be aware of it.

So explicit `--dbpath` overrides value from config file.

rhcarvalho replied:
Hmm, wasn't it the whole point to allow users to pass in whatever config they want? Weird to me to overwrite the dbpath.

Our default config should point to the correct dir, and perhaps we document this concern that you mention for users providing their own custom config file?

omron93 replied:
Yes, it was. And I agree with you.

I would be happier if our scripts would not rely on this. I will try to remove this dependency.

omron93 commented:
Rebased and also added 3.2 version.

So now:

>    Added ability to use user config file.
> - user is able to create or mount custom config file in /etc/mongod.conf (for example from non-docker MongoDB deployments)
> - this config file does not affect MONGODB_REPLICA_NAME (possibility to use user config file is not aimed at replicated environment)

@hhorak @bparees @rhcarvalho If there are no other suggestions please merge.

bparees commented:
lgtm, but please update the READMEs to describe the new supported behavior for supplying a config file:
https://github.com/sclorg/mongodb-container/blob/master/2.4/README.md
https://github.com/sclorg/mongodb-container/blob/master/2.6/README.md
https://github.com/sclorg/mongodb-container/blob/master/3.2/README.md

bparees commented:
[test]

omron93 commented:
@bparees README updated.

Test failed, but it seems that it is not a problem in this PR.

bparees commented:
i'm not sure the failure isn't the fault of this PR:

```
+ docker run --rm openshift/mongodb-24-rhel7-candidate mongo db --host 172.17.0.2 -u user -p password --eval 'db.getSiblingDB('\''test_database'\'');'
exception: connect failed
+ status=1
+ set -e
+ '[' 1 -eq 0 ']'
+ sleep 2
+ echo '  Giving up: Failed to connect. Logs:'
```

seems like it's not able to connect.  are you able to run `make test` on this PR locally successfully?

omron93 commented:
It is fix in this PR, I am sorry.

Should be fixed. Please retest.

bparees commented:
[test]

bparees commented on 2.4/README.md:
```diff
@@ -58,6 +58,14 @@ will be started. If you are re-attaching the volume to another container, the
+
+It is allowed to use custom configuration file for mongod server.
+
+To use custom configuration file in container it has to be mounted into `/etc/mongod.conf`. For example to use configuration file stored in `/home/user` directory use this option for `docker run` command: `-v /home/user/mongod.conf:/etc/mongod.conf:Z`.
```
"Providing a custom configuration file supercedes the individual configuration environment variable values."

omron93 replied:
Look good, thanks. Added.

bparees commented:
one readme nit and lgtm, @hhorak ?

hhorak commented:
@omron93, after 30-upg was merged, could you do the same changes for that version as well, pls?
Otherwise looks good to me, so LGTM.

omron93 commented:
Rebased and added 30-upg.

@bparees @hhorak Please merge.
